nemo/collections/nlp/models/dialogue/intent_bio_type_classification_model.py

```diff
@@ -0,0 +1,821 @@
+            )
+
+            all_generated_slots.append(processed_predicted_slots)
+            all_ground_truth_slots.append(processed_ground_truth_slots)
+            all_utterances.append(' '.join(utterance_tokens))
+
+        os.makedirs(self.cfg.dataset.dialogues_example_dir, exist_ok=True)
+        filename = os.path.join(self.cfg.dataset.dialogues_example_dir, "predictions.jsonl")
+
+        DialogueClassificationMetrics.save_predictions(
+            filename,
+            generated_labels,
+            all_generated_slots,
+            ground_truth_labels,
+            all_ground_truth_slots,
+            ['' for i in range(len(generated_labels))],
+            ['' for i in range(len(generated_labels))],
+            all_utterances,
+        )
+
+        (
+            slot_precision,
+            slot_recall,
+            slot_f1,
+            slot_joint_goal_accuracy,
+        ) = DialogueClassificationMetrics.get_slot_filling_metrics(all_generated_slots, all_ground_truth_slots)
+
+        return slot_precision, slot_recall, slot_f1, slot_joint_goal_accuracy
+
+    @staticmethod
+    def join_tokens_containing_at_sign(utterance_tokens, slot_names):
+        """
+        assumes utterance contains only one @ sign
+        """
+        target_length = len(slot_names)
+        current_length = len(utterance_tokens)
+        diff = current_length - target_length
+        at_sign_positions = [index for index, token in enumerate(utterance_tokens) if token == "@"]
+        try:
+            if len(at_sign_positions) > 1:
+                raise ValueError(
+                    "Current method does not support utterances with more than 1 @ sign ({} encountered), please extend this method for utterance {} with slot names {}".format(
+                        len(at_sign_positions), utterance_tokens, slot_names
+                    )
+                )
+            elif diff == 1:
+                new_tokens = []
+                for index, token in enumerate(utterance_tokens):
+                    if utterance_tokens[index - 1] == "@":
+                        new_tokens[-1] += token
+                    else:
+                        new_tokens.append(token)
+
+            elif diff == 2:
+                new_tokens = []
+                for index, token in enumerate(utterance_tokens[:-1]):
+                    if utterance_tokens[index - 1] == "@" or token == "@":
+                        new_tokens[-1] += token
+                    else:
+                        new_tokens.append(token)
+
+            elif diff == 3:
+                new_tokens = []
+                for index, token in enumerate(utterance_tokens[:-1]):
+                    if utterance_tokens[index + 1] == "@" or utterance_tokens[index - 1] == "@" or token == "@":
+                        new_tokens[-1] += token
+                    else:
+                        new_tokens.append(token)
+            return new_tokens
+        except:
```

Comment:
Is there a more general way to tackle this problem rather than to use try and except?

Reply:
assign @Zhilin123 to solve.

Reply:
solved